compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs

```diff
@@ -143,31 +152,175 @@ pub fn expand_deriving_smart_ptr(
 
     // Find the `#[pointee]` parameter and add an `Unsize<__S>` bound to it.
     let mut impl_generics = generics.clone();
+    let pointee_ty_ident = generics.params[pointee_param_idx].ident;
+    let mut self_bounds;
     {
         let p = &mut impl_generics.params[pointee_param_idx];
+        self_bounds = p.bounds.clone();
         let arg = GenericArg::Type(s_ty.clone());
         let unsize = cx.path_all(span, true, path!(span, core::marker::Unsize), vec![arg]);
         p.bounds.push(cx.trait_bound(unsize, false));
         let mut attrs = thin_vec![];
         swap(&mut p.attrs, &mut attrs);
         p.attrs = attrs.into_iter().filter(|attr| !attr.has_name(sym::pointee)).collect();
     }
+    // We should not set default values to constant generic parameters
+    // and write out bounds that indirectly involves `#[pointee]`.
+    for (idx, (params, orig_params)) in
+        impl_generics.params.iter_mut().zip(&generics.params).enumerate()
+    {
+        if idx == pointee_param_idx {
+            continue;
+        }
+        match &mut params.kind {
+            ast::GenericParamKind::Const { default, .. } => *default = None,
+            ast::GenericParamKind::Type { default } => *default = None,
+            ast::GenericParamKind::Lifetime => {}
+        }
+        for bound in &orig_params.bounds {
+            let mut bound = bound.clone();
+            let mut substitution = TypeSubstitution {
+                from_name: pointee_ty_ident.name,
+                to_ty: &s_ty,
+                rewritten: false,
+            };
+            substitution.visit_param_bound(&mut bound, BoundKind::Bound);
+            if substitution.rewritten {
+                params.bounds.push(bound);
+            }
+        }
+    }
 
     // Add the `__S: ?Sized` extra parameter to the impl block.
+    // We should also write the bounds from `#[pointee]` to `__S` as required by `Unsize<__S>`.
     let sized = cx.path_global(span, path!(span, core::marker::Sized));
-    let bound = GenericBound::Trait(
-        cx.poly_trait_ref(span, sized),
-        TraitBoundModifiers {
-            polarity: ast::BoundPolarity::Maybe(span),
-            constness: ast::BoundConstness::Never,
-            asyncness: ast::BoundAsyncness::Normal,
-        },
-    );
-    let extra_param = cx.typaram(span, Ident::new(sym::__S, span), vec![bound], None);
-    impl_generics.params.push(extra_param);
+    if self_bounds.iter().all(|bound| {
+        if let GenericBound::Trait(
+            trait_ref,
+            TraitBoundModifiers { polarity: ast::BoundPolarity::Maybe(_), .. },
+        ) = bound
+        {
+            !is_sized_marker(&trait_ref.trait_ref.path)
+        } else {
+            false
+        }
```

Comment:
What is this logic even for? Is this ever triggered?

Reply:
Specifically, I'm confused about the combination of `BoundPolarity::Maybe` and `!is_sized_marker`. 

If this is to support `?` for arbitrary traits, we do not support `?Trait` for traits that aren't `Sized`, and I don't think this should account for that.

Reply:
Yes, it is needed.

For some reason, we can repeat `?Sized` bounds on one type twice, be it in generic param position or in where clauses. It is possible that `#[pointee]` has a `?Sized` bound in the `where` clauses of a `struct` definition.